README.md

```diff
@@ -1,2 +1,102 @@
+    1. [Additional Testing Reading](#additional-testing-reading)
+
+## Code Formatting
+
+[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)
+
+We use [`standard`](http://standardjs.com/) in most projects. If it passes `standard`, anything else is up to the developer. No questions asked.
+
+**[back to the top](#table-of-contents)**
+
+## Testing
+
+Let's discuss various tests, what they should do and cover, and how best to organize them.
+
+### Types of Tests
+
+There are three types of tests we should be concerned about.
+
+- Unit
+- Integration
+- Functional
+
+#### Definitions
+
+**Unit** tests are pretty obvious (or should be). They test exactly one unit of functionality. Most likely this should follow the boundaries of functions or at largest, modules. If we're finding we're writing a lot unit tests for a specific function, it may need to be broken up.
+
+An important note: unit tests should _never_ touch an external service (e.g. database).
+
+**Integration** tests make sure that one or more components function correctly together. This can be multiple functions within a module, multiple modules interacting with each other, or a module interacting with a database or external service.
+
+**Functional** tests should align with product ideas. If a user goes through a given flow, they should finish with some outcome. We've found these tests to be pretty difficult to work with as of late, and are trying to get rid of some of these.
+
+#### Organization
+
+In most cases, we should attempt to organize our file structure as follows:
+
+```
+repository/
+  lib/ (or something like it)
+  tests/
+    unit/
+    integration/
+    functional/
+```
+
+Additionally, the `package.json` should support the following commands:
+
+- `npm test`: runs a code linter and all three sets of tests
+- `npm unit`: runs the unit tests
+- `npm integration`: runs the integration tests
+- `npm functional`: runs the functional tests
+
+### Stubbing
+
+One of the most powerful tools we have (and are still learning to use well) is `sinon`. `sinon` is a stubbing and spying library that helps us assert that functions are called with what we expect.
+
+`sinon`'s [docs](http://sinonjs.org/docs/) are quite good, but we will talk about some guidelines to follow when using it:
+
+- In all tests where we are going to stub or spy on something, do so in a `before*`
+- In all tests where we do stub or spy on something,
+    - if it is on an object that is regenerated every test, no action is necessary
+    - if it is on a module or something persistent, always restore the function in an `after*`
+- Stub modules and functions in the correct scopes (meaning, as close to the test that uses it as possible, but hierarchal stubbing is fine)
+- When stubbing functions, we may stub the function to a 'truthy' behavior; doing so reduces code duplication, though the developer will need to adjust the stub(s) to test error cases
```

Comment:
giving flexibility here leads to widely different looking test.
1. setting stub functionality in `it`
```
describe('buzz unit test', function () {
  before(funtion () {
    stub(foo, 'bar')
    stub(foo, 'doo')
  })

  after(function () {
    foo.bar.restore()
    foo.doo.restore()
  })

  // below you would run through all permutations of foo.bar and foo.doo
  it('should fuzz', function() {
    foo.bar.returns(1)
    foo.doo.yieldsAsync('err')

    buzz()
    done()
  })
})
```
pros: in each test it is clear without scrolling anywhere, what each function inside the test should be doing
cons: lots of duplicate code for setting what stub do

2. setting stub functionality in before's
```
describe('buzz unit test', function () {
  before(funtion () {
    stub(foo, 'bar')
    stub(foo, 'doo')
  })

  after(function () {
    foo.bar.restore()
    foo.doo.restore()
  })

  // below you make a describe for all permutations of boo.bar
  describe('foo.bar returns 1', function() {
    before(function(){
      foo.bar.returns(1)
    })

    it('should fuzz', function() {
      foo.doo.yieldsAsync('err')

      buzz()
      done()
    })
  })
})
```
pros: expected stub behavior is explained in the describe block
cons: more cruft (before,describe) code, have to scroll around the block to find out stub behavior